pkg/kallsyms/kallsyms.go

```diff
@@ -127,9 +131,97 @@ func (k *KAllSyms) LookupByInstructionPointer(ip uint64) string {
 }
 
 // SymbolExists returns true if the given symbol exists in the kernel.
-func (k *KAllSyms) SymbolExists(symbol string) bool {
-	_, ok := k.symbolsMap[symbol]
-	return ok
+func SymbolExists(symbol string) bool {
+	_, _, err := KernelSymbolAddress(symbol)
+	return err == nil
+}
+
+// KernelSymbolAddress looks up a symbol from /proc/kallsyms.
+//
+// symbol: the symbol to lookup
+//
+// Returns the address of the symbol and its module if any
+func KernelSymbolAddress(symbol string) (uint64, string, error) {
+	file, err := os.Open("/proc/kallsyms")
+	if err != nil {
+		return 0, "", err
+	}
+	defer file.Close()
+
+	return kernelSymbolAddressFromReader(file, symbol)
+}
+
+// kernelSymbolAddressFromReader
+func kernelSymbolAddressFromReader(r io.Reader, symbol string) (uint64, string, error) {
+	var (
+		count int
+		addr  uint64
+		typ   rune
+		mod   string
+		name  string
+	)
+
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		// Lines looks like:
+		// ffffffff906e97e0 T security_bprm_check
+		// ffffffffc1cb5010 T netem_module_init    [sch_netem]
+
+		line := scanner.Bytes()
+		typeIdx := bytes.IndexByte(line, byte(' '))
+		if typeIdx == -1 || typeIdx+1 >= len(line) {
+			return 0, "", fmt.Errorf("parsing line %q: no type", line)
+		}
+		nameIdx := bytes.IndexByte(line[typeIdx+1:], byte(' '))
+		if nameIdx == -1 || nameIdx+1+nameIdx+1 >= len(line) {
+			return 0, "", fmt.Errorf("parsing line %q: no symbol", line)
+		}
+		modIdx := bytes.IndexByte(line[typeIdx+1+nameIdx+1:], byte(' '))
```

Comment:
Can't we rely on the last byte being `]` to check if the module is present?

Reply:
Yes on current kernel versions but I don't know if future kernels could add more fields.